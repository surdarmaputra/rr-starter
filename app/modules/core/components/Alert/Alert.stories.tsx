import type { Meta, StoryObj } from '@storybook/react';
import { AlertCircle } from 'lucide-react';

import { Alert, AlertDescription, type AlertProps, AlertTitle } from './Alert';

type Story = StoryObj<typeof Alert>;

const meta: Meta = {
  component: Alert,
  title: 'Core/Components/Alert',
};

export default meta;

export const Default: Story = {
  render: (args: AlertProps) => (
    <Alert {...args}>
      <AlertCircle />
      <AlertTitle>Heads up!</AlertTitle>
      <AlertDescription>
        Change a few things up and try submitting again.
      </AlertDescription>
    </Alert>
  ),
};
