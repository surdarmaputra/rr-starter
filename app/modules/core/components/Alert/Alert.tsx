import { cva, type VariantProps } from 'class-variance-authority';
import * as React from 'react';

import { cn } from '~/libs/shadcn/utils';

import type { TestableComponentProps } from '../../types';

const alertVariants = cva(
  'relative w-full rounded-lg border border-slate-200 p-4 [&>svg~*]:pl-8 [&>svg+div]:translate-y-[-3px] [&>svg]:absolute [&>svg]:left-4 [&>svg]:top-4 [&>svg]:text-slate-950 dark:border-slate-800 dark:[&>svg]:text-slate-50',
  {
    variants: {
      variant: {
        default: 'bg-white text-slate-950 dark:bg-slate-950 dark:text-slate-50',
        destructive:
          'border-red-500/50 text-red-500 dark:border-red-500 [&>svg]:text-red-500 dark:border-red-900/50 dark:text-red-900 dark:dark:border-red-900 dark:[&>svg]:text-red-900',
      },
    },
    defaultVariants: {
      variant: 'default',
    },
  },
);

export interface AlertProps
  extends React.HTMLAttributes<HTMLDivElement>,
    VariantProps<typeof alertVariants>,
    TestableComponentProps {}

const Alert = React.forwardRef<HTMLDivElement, AlertProps>(
  ({ className, variant, ...props }, ref) => (
    <div
      className={cn(alertVariants({ variant }), className)}
      ref={ref}
      role="alert"
      {...props}
    />
  ),
);
Alert.displayName = 'Alert';

const AlertTitle = React.forwardRef<
  HTMLParagraphElement,
  React.HTMLAttributes<HTMLHeadingElement>
>(({ className, ...props }, ref) => (
  <h5
    className={cn('mb-1 font-medium leading-none tracking-tight', className)}
    ref={ref}
    {...props}
  />
));
AlertTitle.displayName = 'AlertTitle';

const AlertDescription = React.forwardRef<
  HTMLParagraphElement,
  React.HTMLAttributes<HTMLParagraphElement>
>(({ className, ...props }, ref) => (
  <div
    className={cn('text-sm [&_p]:leading-relaxed', className)}
    ref={ref}
    {...props}
  />
));
AlertDescription.displayName = 'AlertDescription';

export { Alert, AlertDescription, AlertTitle };
