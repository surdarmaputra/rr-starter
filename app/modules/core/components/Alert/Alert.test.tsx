import { render, screen } from '@testing-library/react';
import { describe, expect, test } from 'vitest';

import { Alert, AlertDescription, type AlertProps, AlertTitle } from './Alert';

describe('Alert', () => {
  function renderComponent(props?: AlertProps) {
    return render(
      <Alert {...props}>
        <AlertTitle>Heads up!</AlertTitle>
        <AlertDescription>
          Change a few things up and try submitting again.
        </AlertDescription>
      </Alert>,
    );
  }

  test('renders with data-testid', async () => {
    const dataTestId = 'sample-test-id';
    renderComponent({ 'data-testid': dataTestId });
    await screen.findByTestId(dataTestId);
  });

  test('renders snapshot', async () => {
    const { container } = render(<Alert />);
    renderComponent();
    expect(container).toMatchSnapshot();
  });
});
