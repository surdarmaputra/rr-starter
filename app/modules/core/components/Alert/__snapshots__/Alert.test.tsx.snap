// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Alert > renders snapshot 1`] = `
<div>
  <div
    class="relative w-full rounded-xl p-4 [&>svg~*]:pl-8 [&>svg+div]:translate-y-[-3px] [&>svg]:absolute [&>svg]:left-4 [&>svg]:top-4 bg-accent text-letter-title [&>svg]:text-letter-title"
    role="alert"
  />
</div>
`;
